html/apps/barchart/services/analyticService.js

```diff
@@ -0,0 +1,202 @@
+                
+                
+        }
+        
+        function updateHashtagDateFreq(newTweets, maxHashtags){
+            // console.log("updating hashtagDateFreq");
+
+            // returns an array of obj 
+            // {"hashtag": hashtag, "freq": freq}
+            function getHashtagFreq(tweetArr){
+                
+                // res is an obj of hashtag:count pairs
+                var res = {};
+                
+                var hashtagsArr = tweetArr.map(function(tweet){
+                    return tweet.hashtags;
+                }).reduce(function(prev, next){
+                    return prev.concat(next);
+                }, []);
+                
+                hashtagsArr.forEach(function(hashtag){
+                    res[hashtag] = res[hashtag] || 0;
+                    res[hashtag] ++;
+                })
+                
+                // process res to return array of objects
+                var arr = []; 
+                for (hashtag in res) {
+                    arr.push({
+                        hashtag: hashtag,
+                        frequency: res[hashtag]
+                    });
+                }
+                // console.log(arr);
+                arr.sort(function(a,b){ return b.frequency-a.frequency;});
+                
+                return arr;
+            }
+            
+            function filterTopN(groupedByDate, N){
+                
+                var res = [];
+                // console.log(groupedByDate);
+                groupedByDate.forEach(function(obj){ 
+                    var topNThisDate = {};
+                    // key refers to dayDiff
+                    topNThisDate["day"] = obj["key"];
+                    for(var i=0; i<N; i++){
+                        var topN = "top"+(i+1);                        
+                        topNThisDate[topN] = obj["values"][i] || 0;
+                    }
+                    res.push(topNThisDate);
+                });
+                return res;
+            }
+            
+            var latestDateM =$moment(newTweets[0].created_at).set('hour',0).set('minute',0).set('second',0);
+            
+            // An array containing objects of {"dayDiff": num, "top1": str,.., "topN":str}
+            var groupByDate = d3.nest()
+            // compare by date only, and not time
+            .key(function(d){  
+                var currentM = $moment(d.created_at).set('hour',0).set('minute',0).set('second',0);
+                return latestDateM.diff(currentM, 'days');
+            })
+            // set latest day at idx 0 (as dayDiff = 0)
+            .sortKeys(function(a,b){ return b-a;}) 
+            // process tweets each dayDiff
+            .rollup(function(leaves){ return getHashtagFreq(leaves); })
+            .entries(newTweets);
+            // console.log("grouped", groupByDate);
+            
+            // sort desc freq
+            var filteredFreq = d3.nest()
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: ["d3" is not defined.](https://www.codacy.com/app/sudheesh1995/loklak_server/file/2626095151/issues/source?bid=3288751&fileBranchId=3313532#l192)